spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -112,31 +104,48 @@
     this.sparkContext = JavaSparkContext.fromSparkContext(spark.sparkContext());
     this.table = table;
     this.options = options;
-    this.snapshotId = options.get(SparkReadOptions.SNAPSHOT_ID).map(Long::parseLong).orElse(null);
-    this.asOfTimestamp = options.get(SparkReadOptions.AS_OF_TIMESTAMP).map(Long::parseLong).orElse(null);
-    if (snapshotId != null && asOfTimestamp != null) {
-      throw new IllegalArgumentException(
-          "Cannot scan using both snapshot-id and as-of-timestamp to select the table snapshot");
+
+    // configure a base scan
+    TableScan scan = table.newScan().caseSensitive(caseSensitive);
+
+    Long snapshotId = options.get(SparkReadOptions.SNAPSHOT_ID).map(Long::parseLong).orElse(null);
+    if (snapshotId != null) {
+      scan = scan.useSnapshot(snapshotId);
     }
 
-    this.startSnapshotId = options.get("start-snapshot-id").map(Long::parseLong).orElse(null);
-    this.endSnapshotId = options.get("end-snapshot-id").map(Long::parseLong).orElse(null);
-    if (snapshotId != null || asOfTimestamp != null) {
-      if (startSnapshotId != null || endSnapshotId != null) {
-        throw new IllegalArgumentException(
-            "Cannot specify start-snapshot-id and end-snapshot-id to do incremental scan when either snapshot-id or " +
-                "as-of-timestamp is specified");
-      }
-    } else {
-      if (startSnapshotId == null && endSnapshotId != null) {
-        throw new IllegalArgumentException("Cannot only specify option end-snapshot-id to do incremental scan");
```

Comment:
These checks are tested in `TestDataSourceOptions.testIncrementalScanOptions`. We need to retain them (somewhere).

Reply:
These are checked in the table scan itself rather than in multiple places. I think we just need to update the tests to match the error messages from the scan.

Reply:
Let us move this cleanup (of checks in multiple places) to a separate PR after https://github.com/apache/iceberg/pull/1508 is merged. It needs to be done for spark3 as well, and is really orthogonal to the main issue.